File: M3Tools/Types/People/Listener.vb
Option Strict On

Namespace Types
	Public Class Listener
		Inherits Person

		<Text.Json.Serialization.JsonPropertyName("listenerID")>
		Overrides Property Id As Integer

		Public Sub New()
			Me.New(-1)
		End Sub

		Public Sub New(listenerID As Integer, Optional fName As String = "Test", Optional lName As String = "Listener", Optional email As String = "TestListener@domain.tst")
			Me.New(listenerID, $"{fName} {lName}", email)
		End Sub

		Public Sub New(listenerID As Integer, name As String, email As String)
			MyBase.New(listenerID, name, email)
		End Sub

		Public Overrides Function ToString() As String
			Return $"{Id}) {Name} <{Email}>"
		End Function

		Shared Function Parse(arr As Object()) As Listener
			Return New Listener(CInt(arr(0)), CStr(arr(1)), CStr(arr(2)))
		End Function

		Public Function Clone() As Listener
			Return New Listener(Me.Id, Me.Name, Me.Email)
		End Function
	End Class
End Namespace

File: M3Tools/Types/People/Listener.cs
﻿namespace SPPBC.M3Tools.Types
{
	/// <summary>
	/// An email listener that receives the email ministry emails
	/// </summary>
	[Newtonsoft.Json.JsonObject]
	public class Listener : Person
	{
		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		[Newtonsoft.Json.JsonProperty("listenerID")]
		public override int Id
		{
			get => base.Id;
			set => base.Id = value;
		}

		/// <summary>
		/// An empty instance of the Listener class
		/// </summary>
		public static new Listener None => new();

		/// <summary>
		/// 
		/// </summary>
		public Listener() : this(-1)
		{
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="listenerID"></param>
		/// <param name="fName"></param>
		/// <param name="lName"></param>
		/// <param name="email"></param>
		[Newtonsoft.Json.JsonConstructor]
		public Listener(int listenerID, string fName = "New", string lName = "Listener", string email = default) : this(listenerID, $"{fName} {lName}", email)
		{
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="listenerID"></param>
		/// <param name="name"></param>
		/// <param name="email"></param>
		public Listener(int listenerID, string name, string email) : base(listenerID, name, email)
		{
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <returns></returns>
		public override string ToString() => $"{Id}) {Name} <{Email}>";

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		/// <returns></returns>
		public static bool operator ==(Listener left, Listener right) => !(left is null ^ right is null) && left.Id == right.Id && left.Name == right.Name && left.Email == right.Email;

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		/// <returns></returns>
		public static bool operator !=(Listener left, Listener right) => !(left == right);

		/// <inheritdoc/>
		public override bool Equals(object obj) => this == (obj as Listener);

		/// <inheritdoc/>
		public override int GetHashCode() => base.GetHashCode();
	}
}
